src/adapters/WindowAdapter.ts

```diff
@@ -0,0 +1,76 @@
+export default class WindowAdapter {
+    controlledWindow?: Window;
+
+    open: boolean;
+
+    onClose?: Function;
+
+    get isInitialized() {
+        return this.controlledWindow !== undefined && !this.controlledWindow.closed;
+    }
+
+    constructor(onClose?: Function) {
+        this.open = false;
+        this.onClose = onClose;
+    }
+
+    async init({
+        parentWindow,
+        url,
+        width = 375,
+        height = 650,
+        target = "popupWindow",
+    }: {
+        parentWindow: Window;
+        url: string;
+        width?: number;
+        height?: number;
+        target?: string;
+    }) {
+        return await new Promise<Window | undefined>((resolve, reject) => {
+            const _window = parentWindow.open(url, target, this.createPopupString(width, height));
+
+            if (!_window) {
+                reject(new Error(`Failed to open popup. This may be caused by the browsers' popup blocker`));
+            }
+
+            this.controlledWindow = _window ?? undefined;
+            this.open = true;
+
+            this.registerListeners();
+            resolve(this.controlledWindow);
+        });
+    }
+
+    private registerListeners() {
+        const timer = setInterval(() => {
+            if (this.controlledWindow?.closed) {
+                clearInterval(timer);
+
+                this.controlledWindow = undefined;
+                this.open = false;
+
+                if (this.onClose) {
+                    this.onClose();
+                }
+            }
+        }, 50);
+    }
```

Comment:
isn't there some way to add an actual listener instead of this active polling?

Reply:
unfortunately no :(